pkg/restic/backupper_restorer.go

```diff
@@ -0,0 +1,286 @@
+	Restorer
+}
+
+// Backupper can execute restic backups of volumes in a pod.
+type Backupper interface {
+	BackupPodVolumes(backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+// Restorer can execute restic restores of volumes in a pod.
+type Restorer interface {
+	RestorePodVolumes(restore *arkv1api.Restore, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+func (br *backupperRestorer) BackupPodVolumes(backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error {
+	// get volumes to backup from pod's annotations
+	volumesToBackup := GetVolumesToBackup(pod)
+	if len(volumesToBackup) == 0 {
+		return nil
+	}
+
+	// Get existing snapshots annotated on backup, and fail-fast if
+	// we can't. We'll append new snapshots to this at the end of
+	// this func.
+	backupSnapshots, err := GetSnapshotsInBackup(backup)
+	if err != nil {
+		return err
+	}
+
+	// ensure a repo exists for the pod's namespace
+	exists, err := br.metadataManager.RepositoryExists(pod.Namespace)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if err := br.metadataManager.InitRepo(pod.Namespace); err != nil {
+			return err
+		}
+	}
+
+	var (
+		errs []error
+		lock sync.Mutex
+		wg   sync.WaitGroup
+	)
+
+	// for each volume to backup:
+	for _, volumeName := range volumesToBackup {
+		wg.Add(1)
+		go func(volumeName string) {
+			defer wg.Done()
+
+			volume := getVolume(pod, volumeName)
+			if volume == nil {
+				lock.Lock()
+				errs = append(errs, errors.Errorf("volume %s does not exist in pod %s", volumeName, kube.NamespaceAndName(pod)))
+				lock.Unlock()
+				return
+			}
+
+			// get the volume's directory name under /var/lib/kubelet/pods/... on the host
+			volumeDir, err := getVolumeDirectory(volume, pod.Namespace, br.pvcGetter)
+			if err != nil {
+				lock.Lock()
+				errs = append(errs, err)
+				lock.Unlock()
+				return
+			}
+
+			// assemble restic backup command
+			snapshotTags := map[string]string{
+				"backup":     backup.Name,
+				"backup-uid": string(backup.UID),
+				"pod":        pod.Name,
+				"pod-uid":    string(pod.UID),
+				"ns":         pod.Namespace,
+				"volume":     volumeName,
+			}
+
+			cmd := backupCommand(br.metadataManager.RepoPrefix(), pod.Namespace, string(pod.UID), volumeDir, snapshotTags)
+
+			// exec it in the daemonset pod
+			if err := br.daemonSetExecutor.Exec(pod.Spec.NodeName, cmd, 10*time.Minute, log); err != nil {
+				lock.Lock()
+				errs = append(errs, err)
+				lock.Unlock()
+				return
+			}
+
+			// get the snapshot's ID
+			snapshotID, err := br.metadataManager.GetSnapshotID(pod.Namespace, string(backup.UID), string(pod.UID), volumeName)
+			if err != nil {
+				lock.Lock()
+				errs = append(errs, err)
+				lock.Unlock()
+				return
+			}
+
+			// save the snapshot's ID in the pod annotation
+			lock.Lock()
+			SetPodSnapshotAnnotation(pod, volumeName, snapshotID)
+			backupSnapshots = append(backupSnapshots, fmt.Sprintf("%s/%s", pod.Namespace, snapshotID))
```

Comment:
Instead of using a lock, could we somehow use channel(s) to coordinate retrieving the snapshot id?